README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
![coreSim logo](img/coreSim_logo.png)

Version: 0.2.4 [![CRAN
Version](http://www.r-pkg.org/badges/version/coreSim)](http://cran.r-project.org/package=coreSim)
[![Build
Status](https://travis-ci.org/christophergandrud/coreSim.svg?branch=master)](https://travis-ci.org/christophergandrud/coreSim)
[![codecov.io](https://codecov.io/github/christophergandrud/coreSim/coverage.svg?branch=master)](https://codecov.io/github/christophergandrud/coreSim?branch=master)
![CRAN Monthly
Downloads](http://cranlogs.r-pkg.org/badges/last-month/coreSim) ![CRAN
Total Downloads](http://cranlogs.r-pkg.org/badges/grand-total/coreSim)

> Core functionality for simulating quantities of interest from
> generalised linear models.

Purpose
=======

Using simulations to find quantities of interest and associated
uncertainty can be an effective way of showing substantively meaningful
results from generalised linear models (GLM).

This R package provides core functions that can serve as the
**backbone** to other packages for finding and plotting simulated
quantities of interest from GLMs.

**coreSim** currently powers
[pltesim](https://github.com/christophergandrud/pltesim), a package for
simulating probabilistic long-term effects from models with temporal
dependence.

Motivation
==========

**coreSim** aims to solve a number of issues that arose in prior
implementations of the simulation approach to showing GLM results. The
main previous implementation in R is the **Zelig** package. This package
has tried to be "Everyone's statistical software". However,
paradoxically, its attempt to be everything to everyone has led to less
flexibility for new use cases. Maintaining such a large project over
time has led to (in my experience) frequent code breaks. The Zelig 'API'
has changed considerably over time in often undocumented ways. Changes
to its many dependencies also undermines its reliability.

**coreSim** aims to overcome these issue with a focus on *simplicity*.
It tries to:

-   Do a small set of things really well.

-   Have as few dependencies as possible. Only import packages if they
    make significant performance improvements over base R.

-   Return simple `data.frame` output that can be easily manipulated.

-   Have informative error messages that are easy for users to
    understand and which guide them to solutions.

Additionally, **coreSim** aims for very high reliability. Simplicity
helps with acheive this goal. So does aiming for 100% code test
coverage.

These characteristics allow **coreSim** to form the backbone of many
specific and unanticipated implementations of the simulation approach.

 Steps
======

1.  Estimate your model using whatever GLM model fitting function you
    like (note: I've only tested `lm`, `glm`, and `survival`).

2.  Find your quantities of interest with `qi_builder`.

3.  Present your results, e.g. by plotting the simulated quantities of
    interest.

Examples
========

Normal linear model
-------------------

Here is an example using data from the **car** package:

    library(coreSim)
    library(car)

    # Normal linear model
    m1 <- lm(prestige ~ education + type, data = Prestige)

    # Create fitted values
    fitted_df_1 <- expand.grid(education = 6:16, type = 'wc')

    # Find predicted outcomes (95% central interval, by default)
    linear_qi <- qi_builder(obj = m1, newdata = fitted_df_1)

    ## Note: FUN argument missing -> assuming b_sims is from a normal linear model.

    head(linear_qi)

    ##   education typewc      qi_
    ## 1         6      1 18.26931
    ## 2         6      1 20.92067
    ## 3         6      1 20.88659
    ## 4         6      1 24.46231
    ## 5         6      1 19.15713
    ## 6         6      1 17.91391

### Slimmed simulation data

By default `qi_builder` will return all of the simulations inside the
central interval of the simulations for each scenario that you specify
with the `ci` argument (this is `0.95` by default for 95% central
interval).

However, you may want to only return key features of this interval so
that they can be efficiently stored and plotted. Using `slim = TRUE`
will return only the minimum, median, and maximum values of the central
interval for each scenario:

    linear_qi_slim <- qi_builder(m1, newdata = fitted_df_1, slim = TRUE)

    head(linear_qi_slim)

    ##   education typewc   qi_min qi_median   qi_max
    ## 1         6      1 12.54993  19.33749 26.46225
    ## 2         7      1 18.10320  23.88602 29.90101
    ## 3         8      1 23.71457  28.48066 33.52652
    ## 4         9      1 29.04148  33.05269 36.90274
    ## 5        10      1 34.09772  37.64164 41.08075
    ## 6        11      1 38.99077  42.18919 45.31262

The slimmed simulation data set can be efficiently plotted, for example
using [ggplot2](http://docs.ggplot2.org/current/):

    library(ggplot2)
    theme_set(theme_bw())

    ggplot(linear_qi_slim, aes(education, qi_median)) +
        geom_ribbon(aes(ymin = qi_min, ymax = qi_max), alpha = 0.3) +
        geom_line() +
        ylab('Prestige')

![](man/figures/unnamed-chunk-3-1.png)

Predicted probabilities from logistic regressions
-------------------------------------------------

By default `qi_builder` simply returns the linear systematic component,
which in normal linear regression is simply the predicted y, i.e.
*y* = *α* + *β**X*.

`qi_builder` allows you to supply any function for creating quantities
of interest that you would like. This function needs to simply be able
to convert a vector of linear systematic components to your quantity of
interest.

For example, to find predicted probabilities from a logistic regression
model create a function to turn the systematic component into the QI:

    pr_fun <- function(x) 1 / (1 + exp(-x))

Then supply the custom function to `qi_builder`'s `FUN` argument:

    # Load data
    data(Admission)
    Admission$rank <- as.factor(Admission$rank)

    # Estimate model
    m2 <- glm(admit ~ gre + gpa + rank, data = Admission, family = 'binomial')

    # Create fitted values
    m2_fitted <- expand.grid(gre = seq(220, 800, by = 10), gpa = c(1, 4), rank = '4')

    # Find quantity of interest
    logistic_qi <- qi_builder(m2, m2_fitted, FUN = pr_fun, slim = TRUE)

    head(logistic_qi)

    ##   gre gpa rank4      qi_min  qi_median     qi_max
    ## 1 220   1     1 0.002879333 0.01431348 0.06336300
    ## 2 230   1     1 0.002983825 0.01471619 0.06436528
    ## 3 240   1     1 0.003106020 0.01501327 0.06449270
    ## 4 250   1     1 0.003205407 0.01537888 0.06520055
    ## 5 260   1     1 0.003301195 0.01566664 0.06620142
    ## 6 270   1     1 0.003415190 0.01588294 0.06787158

See also
========

Christopher Gandrud (2015). simPH: An R Package for Illustrating
Estimates from Cox Proportional Hazard Models Including for Interactive
and Nonlinear Effects. Journal of Statistical Software, 65(3), 1-20.
<http://www.jstatsoft.org/v65/i03/>.

Gandrud, Christopher. Laron K. Williams and Guy D. Whitten (2015).
dynsim: Dynamic Simulations of Autoregressive Relationships. R package
version 1.2.1. <https://CRAN.R-project.org/package=dynsim>.

King, Gary, Michael Tomz, and Jason Wittenberg. 2000. "Making the Most
of Statistical Analyses: Improving Interpretation and Presentation."
American Journal of Political Science 44(2): 341-55.


NEWS
# Version 0.2.4

- Admissions example data now included with package to avoid breaking URL.

# Version 0.2.3

- Internal cleanup of function tests.

# Version 0.2.2

- Allow users to supply mu and Sigma values to `b_sim`.

- Documentation and internal code improvements.

# Version 0.2.1

- Can find polynomial splines using b-splines basis. Note: apart from the
degree of the piecewise polynomial, the default behaviour for `bs` is
assumed.

- Internal code improvements.

# Version 0.2

- `b_sim` has been incorporated into qi_builder so that the simulated
coefficients and QI calculations can be accomplished in one function call.

- Handles categorical variables entered as `newdata` either in their
original format or matching the coefficient name output.

- Can find polynomially transformed terms if entered in the model fit function
with there base term + the `I` function. Note, currently only supports up to
`x^9`.

- Can find natural logarithmically transfromed terms if entered in the model
fit function with the `log` function.

- Added `original_order` to `qi_builder`. If `TRUE` then the values are returned
in the original simulation order. Only relevant if `slim = TRUE`.


# Version 0.1

- Minimal working version.


tests/testthat.R
library(testthat)
library(coreSim)
library(car)
library(survival)

test_check("coreSim", report = 'summary')


tests/testthat/test.R
# Test b_sim -------------------------------------------------------------------
test_that('b_sim output validity', {
    set.seed(100)

    # Linear model
    data('Prestige', package = 'car')
    m1 <- lm(prestige ~ education + type, data = Prestige)
    m1_sims <- b_sim(m1)

    # Manually supplied coefficient means and covariance matrix
    coefs <- coef(m1)
    vcov_matrix <- vcov(m1)
    prestige_sims_manual <- b_sim(mu = coefs, Sigma = vcov_matrix)

    # Preference for obj

    # Survival model (no intercept term)
    library(survival)
    test1 <- list(time = c(4,3,1,1,2,2,3),
                  status = c(1,1,1,0,1,1,0),
                  x = c(0,2,1,1,1,0,0),
                  sex = c(0,0,0,0,1,1,1))
    # Fit a stratified model
    m_coxph <- coxph(Surv(time, status) ~ x + strata(sex), test1)
    m_coxph <- sim_coxph <- b_sim(m_coxph)

    expect_match(names(m1_sims)[[1]], 'intercept_')
    expect_match(names(prestige_sims_manual)[[1]], 'intercept_')
    expect_message(b_sim(obj = m1, mu = coefs, Sigma = vcov_matrix),
                   'obj used rather than mu and/or Sigma')
    expect_match(names(sim_coxph)[[1]], 'x')
})

# Test linear_systematic -------------------------------------------------------
test_that('linear_systematic output validity', {
    set.seed(100)

    library(splines)

    # Linear model
    data('Prestige', package = 'car')
    m1 <- lm(prestige ~ education + type, data = Prestige)
    m1_sims <- b_sim(m1)

    fitted_df <- expand.grid(education = 6:16, typewc = 1)
    ls_lm <- linear_systematic(b_sims = m1_sims, newdata = fitted_df)
    ls_lm_no_intercept <- linear_systematic(b_sims = m1_sims,
                                            newdata = fitted_df,
                                            inc_intercept = FALSE)

    mpoly <- lm(prestige ~ education + I(education ^2) + type, data = Prestige)
    mpoly_sims <- b_sim(mpoly)
    ls_mpoly <- linear_systematic(b_sims = mpoly_sims, newdata = fitted_df)

    mlog <- lm(prestige ~ log(education) + type, data = Prestige)
    mlog_sims <- b_sim(mlog)
    ls_mlog <- linear_systematic(b_sims = mlog_sims, newdata = fitted_df)

    mbs1 <- lm(prestige ~ bs(education) + type, data = Prestige)
    mbs1_sims <- b_sim(mbs1)
    ls_mbs1 <- linear_systematic(b_sims = mbs1_sims, newdata = fitted_df)

    mbs2 <- lm(prestige ~ bs(education, degree = 3) + type, data = Prestige)
    mbs2_sims <- b_sim(mbs2)
    ls_mbs2 <- linear_systematic(b_sims = mbs2_sims, newdata = fitted_df)

    # Survival model (no intercept term)
    library(survival)
    test1 <- list(time = c(4,3,1,1,2,2,3),
                  status = c(1,1,1,0,1,1,0),
                  x = c(0,2,1,1,1,0,0),
                  sex = c(0,0,0,0,1,1,1))
    # Fit a stratified model
    m_coxph <- coxph(Surv(time, status) ~ x + strata(sex), test1)
    sim_coxph <- b_sim(m_coxph)
    ls_coxph <- linear_systematic(b_sims = sim_coxph,
                                  newdata = data.frame(x = 1))

    expect_is(ls_lm$ls_, 'numeric')
    expect_true(mean(ls_lm$ls_) < mean(ls_lm_no_intercept$ls_))
    expect_is(ls_coxph$ls_, 'numeric')
    expect_equal(names(ls_mpoly), c('education', 'typewc', 'I.education.2.',
                                    'ls_'))
    expect_equal(names(ls_mlog), c('typewc', 'log.education.', 'ls_'))
    expect_equal(names(ls_mbs1), c('typewc', 'bs.education.1', 'bs.education.2',
                                  'bs.education.3', 'ls_' ))
    expect_equal(names(ls_mbs2), c('typewc', 'bs.education..degree...3.1',
                                   'bs.education..degree...3.2',
                                   'bs.education..degree...3.3', 'ls_' ))
})

# Test qi_builder --------------------------------------------------------------
test_that('qi_builder output validity', {
    set.seed(100)

    # Linear model
    data('Prestige', package = 'car')
    m1 <- lm(prestige ~ education + type, data = Prestige)
    m1_sims <- b_sim(m1)

    fitted_df <- expand.grid(education = c(7, 6, 8:16), typewc = 1)

    linear_qi <- qi_builder(b_sims = m1_sims, newdata = fitted_df)
    linear_qi_c1 <- qi_builder(m1, newdata = fitted_df, ci = 1)
    linear_qi_slim1 <- qi_builder(m1, newdata = fitted_df, slim = TRUE)
    linear_qi_slim2 <- qi_builder(m1, newdata = fitted_df,
                                  original_order = TRUE, slim = TRUE)

    linear_qi_auto_newdata <- qi_builder(m1)

    # Manually supply coefficient means and covariance matrix
    coefs <- coef(m1)
    vcov_matrix <- vcov(m1)

    linear_qi_custom_mu_Sigma <- qi_builder(mu = coefs, Sigma = vcov_matrix,
                                            newdata = fitted_df)

    # Predicted probabilities from logistic regression
    URL <- 'http://www.ats.ucla.edu/stat/data/binary.csv'
    Admission <- read.csv(URL)
    Admission$rank <- as.factor(Admission$rank)
    m2 <- glm(admit ~ gre + gpa + rank, data = Admission, family = 'binomial')

    m2_sims <- b_sim(m2)

    m2_fitted <- expand.grid(gre = seq(220, 800, by = 10), gpa = c(2, 4),
                             rank4 = 1)
    pr_function <- function(x) 1 / (1 + exp(-x))
    logistic_qi <- qi_builder(m2, m2_fitted, FUN = pr_function)

    expect_is(linear_qi$qi_, 'numeric')
    expect_is(logistic_qi$qi_, 'numeric')
    expect_equal(nrow(linear_qi_c1), 11000)
    expect_equal(nrow(linear_qi_slim1), 11)
    expect_equal(names(linear_qi_slim1), c('education', 'typewc', 'qi_min',
                                       'qi_median', 'qi_max'))
    expect_equal(nrow(linear_qi_auto_newdata), 88350)
    expect_false(all(linear_qi_slim1$education == linear_qi_slim2$education))
    expect_equal(nrow(linear_qi_custom_mu_Sigma), nrow(linear_qi))
    expect_error(qi_builder(m1, nsim = 20000))
    expect_error(qi_builder(m2, m2_fitted, FUN = pr_function, ci = 950))
    expect_error(qi_builder(m2, m2_fitted, FUN = function(x, y){x + y}))
    expect_error(qi_builder(m2, m2_fitted, FUN = function(x){x <- 'a'}))
    expect_error(qi_builder(m2, m2_fitted, FUN = function(x){x <- data.frame()}))
    expect_error(qi_builder(m2, m2_fitted, FUN = 'test_fail'))

    # Unsupported model for automatic newdata creation
    library(survival)
    test1 <- list(time = c(4,3,1,1,2,2,3),
                  status = c(1,1,1,0,1,1,0),
                  x = c(0,2,1,1,1,0,0),
                  sex = c(0,0,0,0,1,1,1))
    # Fit a stratified model
    m_coxph <- coxph(Surv(time, status) ~ x + strata(sex), test1)
    expect_error(qi_builder(m_coxph))

})

# Test qi_slimmer --------------------------------------------------------------
test_that('qi_slimmer output validity', {
    set.seed(100)

    # Linear model
    data('Prestige', package = 'car')
    m1 <- lm(prestige ~ education + type, data = Prestige)
    m1_sims <- b_sim(m1)

    fitted_df <- expand.grid(education = 6:16, type = 'wc')
    linear_qi <- qi_builder(b_sims = m1_sims, newdata = fitted_df)

    linear_slim <- qi_slimmer(linear_qi)

    expect_equal(nrow(linear_slim), 11)
    expect_equal(names(linear_slim), c('education', 'typewc', 'qi_min',
                                       'qi_median', 'qi_max'))
})


# Test interaction_builder -----------------------------------------------------
test_that('interaction_builder output validity', {
    states <- as.data.frame(state.x77)
    states$in.come <- states$Income
    m1 <- lm(Murder ~ Illiteracy + Income * Population, data = states)
    b_sims_1 <- b_sim(m1)

    m2 <- lm(Murder ~ in.come * Population, data = states)
    b_sims_2 <- b_sim(m2)

    m3 <- lm(Murder ~ in.come * Population * Frost, data = states)
    b_sims_3 <- b_sim(m3)

    fitted_df_1 <- expand.grid(Income = unique(states$Income), Illiteracy = 1,
                               Population = 4246)

    fitted_df_2 <- fitted_df_1
    fitted_df_2$Income.Population <- fitted_df_2$Income * fitted_df_2$Population

    fitted_df_3 <- expand.grid(in.come = unique(states$in.come),
                               Population = 4246, Frost = 104)

    inter_df_1 <- coreSim:::interaction_builder(b_sims_1, fitted_df_1)
    inter_df_2 <- coreSim:::interaction_builder(b_sims_1, fitted_df_2)
    inter_df_3 <- coreSim:::interaction_builder(b_sims_2, fitted_df_3)
    inter_df_4 <- coreSim:::interaction_builder(b_sims_3, fitted_df_3)

    inter_systematic <- linear_systematic(b_sims_1, fitted_df_1)

    expect_equal(inter_df_1$Income * inter_df_1$Population,
                 inter_df_1$Income.Population)
    expect_equal(inter_df_1$Income.Population, inter_df_2$Income.Population)
    expect_equal(inter_df_1$Income.Population, inter_df_3$in.come.Population)
    expect_equal(names(inter_df_4), c('in.come', 'Population', 'Frost',
                                      'in.come.Population', 'in.come.Frost',
                                      'Population.Frost',
                                      'in.come.Population.Frost'))
    expect_equal(ncol(inter_systematic), 5)


    fake_names <- c('a', 'b', 'c', 'd', 'e')
    interaction_5 <- coreSim:::possible_interaction_terms(fake_names, n = 5)

    expect_equal(length(interaction_5), 120)
    expect_error(coreSim:::possible_interaction_terms(fake_names, n = 6))
})

# Test ci_check --------------------------------------------------------------
test_that('ci_check output validity', {
    expect_equal(coreSim:::ci_check(95), 0.95)
    expect_equal(coreSim:::ci_check(1), 1)
    expect_equal(coreSim:::ci_check(0.5), 0.5)
    expect_error(coreSim:::ci_check(500))
})


tests/testthat/test_bsim.R
context('b_sim')

test_that('b_sim output validity', {
    set.seed(100)

    # Linear model
    data('Prestige', package = 'car')
    m1 <- lm(prestige ~ education + type, data = Prestige)
    m1_sims <- b_sim(m1)

    # Manually supplied coefficient means and covariance matrix
    coefs <- coef(m1)
    vcov_matrix <- vcov(m1)
    prestige_sims_manual <- b_sim(mu = coefs, Sigma = vcov_matrix)

    # Preference for obj

    # Survival model (no intercept term)
    library(survival)
    test1 <- list(time = c(4,3,1,1,2,2,3),
                  status = c(1,1,1,0,1,1,0),
                  x = c(0,2,1,1,1,0,0),
                  sex = c(0,0,0,0,1,1,1))
    # Fit a stratified model
    m_coxph <- coxph(Surv(time, status) ~ x + strata(sex), test1)
    m_coxph <- sim_coxph <- b_sim(m_coxph)

    expect_match(names(m1_sims)[[1]], 'intercept_')
    expect_match(names(prestige_sims_manual)[[1]], 'intercept_')
    expect_message(b_sim(obj = m1, mu = coefs, Sigma = vcov_matrix),
                   'obj used rather than mu and/or Sigma')
    expect_match(names(sim_coxph)[[1]], 'x')
})


R/b_sim.R
#' Simulate coefficients from a GLM by making draws from the multivariate
#' normal distribution
#'
#' @param obj a fitted model object.
#' @param mu an optional vector giving the means of the variables. If \code{obj}
#' is supplied then \code{mu} is ignored.
#' @param Sigma an optional positive-definite symmetric matrix specifying the
#' covariance matrix of the variables. If \code{obj} is supplied then
#' \code{Sigma} is ignored. If your model includes an intercept, this should be
#' given the name \code{intercept_}.
#' @param nsim number of simulations to draw.
#'
#' @return A data frame of simulated coefficients from \code{obj}.
#'
#' @examples
#' library(car)
#'
#' # Estimate model
#' m1 <- lm(prestige ~ education + type, data = Prestige)
#'
#' # Create fitted values
#' prestige_sims <- b_sim(m1)
#'
#' # Manually supply coefficient means and covariance matrix
#' coefs <- coef(m1)
#' vcov_matrix <- vcov(m1)
#'
#' prestige_sims_manual <- b_sim(mu = coefs, Sigma = vcov_matrix)
#'
#' @importFrom stats coef vcov
#' @importFrom MASS mvrnorm
#'
#' @export

b_sim <- function(obj, mu, Sigma, nsim = 1000)
{
    if (!missing(obj)) {
        if (!missing(mu) || !missing(Sigma))
            message('obj used rather than mu and/or Sigma')
        obj_coef <- coef(obj)
        obj_vcov <- vcov(obj)
    }
    else if (missing(obj)) {
        if (!is.vector(mu)) stop('mu must be a vector', call. = FALSE)
        if (!is.matrix(Sigma)) stop('Sigma must be a matrix', call. = FALSE)
        if (length(mu) != nrow(Sigma) | length(mu) != ncol(Sigma))
            stop("The length of mu and Sigma's rows and columns must be equal.",
                 call. = FALSE)

        obj_coef <- mu
        obj_vcov <- Sigma
    }

    drawn <- mvrnorm(n = nsim, mu = obj_coef, Sigma = obj_vcov)
    drawn <- data.frame(drawn)

    if (missing (obj)) names(drawn) <- colnames(obj_vcov)

    if (grepl('intercept', names(drawn[1]), ignore.case = TRUE))
        names(drawn)[1] <- 'intercept_'
    return(drawn)
}


R/qi_builder.R
#' Find quantities of interest from generalized linear models
#'
#' @param obj a fitted model object from which to base coefficient simulations
#' on.
#' @param newdata an optional data frame of fitted values with column names
#' corresponding to coefficient names in \code{obj} or \code{mu}/\code{Sigma}.
#' Note that variables not included in \code{newdata} will be fitted at 0. If
#' \code{missing} then observations used to fit the model in \code{obj} will be
#' used.
#' @param FUN a function for calculating how to find the quantity of interest
#' from a vector of the fitted linear systematic component. It must return
#' a numeric vector. If \code{missing} then a normal
#' linear regression model is assumed and the predicted values are returned
#' (i.e. the fitted linear systematic component from
#' \code{\link{linear_systematic}}).
#' @param nsim number of simulations to draw.
#' @param ci the proportion of the central interval of the simulations to
#' return. Must be in (0, 1] or equivalently (0, 100]. Note: if \code{ci = 1}
#' then the full interval (i.e. 100 percent) is assumed.
#' @param slim logical indicating whether to (if \code{FALSE}) return all
#' simulations in the central interval specified by \code{ci} for each fitted
#' scenario or (if \code{TRUE}) just the minimum, median, and maxium values.
#' See \code{\link{qi_slimmer}} for more details.
#' @param large_computation logical. If \code{newdata} is not supplied,
#' whether to allow > 100000 simulated quantities of interest to be found.
#' @param original_order logical whether or not to keep the original scenario
#' order when \code{slim = TRUE}. Choosing \code{FALSE} can imporove computation
#' time.
#' @param b_sims an optional data frame created by \code{\link{b_sim}} of
#' simulated coefficients. Only used if \code{obj} is not supplied.
#' @param mu an optional vector giving the means of the variables. If \code{obj}
#' or \code{b_sims} is supplied then \code{mu} is ignored.
#' @param Sigma an optional positive-definite symmetric matrix specifying the
#' covariance matrix of the variables. If \code{obj} is supplied then
#' \code{Sigma} is ignored. If your model includes an intercept, this should be
#' given the name \code{intercept_}.
#' @param verbose logical. Whether to include full set of messages or not.
#' @param ... arguments to passed to \code{\link{linear_systematic}}.
#'
#' @return If \code{slimmer = FALSE} a data frame of fitted values supplied in
#' \code{newdata} and associated simulated quantities of interest for all
#' simulations in the central interval specified by \code{ci}. The quantities
#' of interest are in a column named \code{qi_}.
#'
#' If \code{slimmer = TRUE} a data frame of fitted values supplied in
#' \code{newdata} and the minimum, median, and maximum values of the central
#' interval specified by \code{ci} for each scenario are returned in three
#' columns named \code{qi_min}, \code{qi_median}, and \code{qi_max},
#' respectively.
#'
#' @examples
#' library(car)
#'
#' ## Normal linear model
#' m1 <- lm(prestige ~ education + type, data = Prestige)
#'
#' # Using observed data as scenarios
#' linear_qi_obs <- qi_builder(m1)
#'
#' # Create fitted values
#' fitted_df_1 <- expand.grid(education = 6:16, typewc = 1)
#'
#' linear_qi <- qi_builder(m1, newdata = fitted_df_1)
#'
#' # Manually supply coefficient means and covariance matrix
#' coefs <- coef(m1)
#' vcov_matrix <- vcov(m1)
#'
#' linear_qi_custom_mu_Sigma <- qi_builder(mu = coefs, Sigma = vcov_matrix,
#'                                  newdata = fitted_df_1)
#'
#' ## Logistic regression
#' # Load data
#' data(Admission)
#' Admission$rank <- as.factor(Admission$rank)
#'
#' # Estimate model
#' m2 <- glm(admit ~ gre + gpa + rank, data = Admission, family = 'binomial')
#'
#' # Specify fitted values
#' m2_fitted <- expand.grid(gre = seq(220, 800, by = 10), gpa = c(2, 4),
#'                          rank = '4')
#'
#' # Function to find predicted probabilities from logistic regression models
#' pr_function <- function(x) 1 / (1 + exp(-x))
#'
#' # Find quantity of interest
#' logistic_qi_1 <- qi_builder(m2, m2_fitted, FUN = pr_function)
#'
#' logistic_qi_2 <- qi_builder(m2, m2_fitted, FUN = pr_function,
#'                          slim = TRUE)
#'
#' @importFrom dplyr bind_rows
#'
#' @export

qi_builder <- function(obj, newdata, FUN, ci = 0.95, nsim = 1000,
                      slim = FALSE,  large_computation = FALSE,
                      original_order = FALSE,
                      b_sims, mu, Sigma, verbose = TRUE,
                      ...)
{
    qi_ <- NULL

    ci <- ci_check(ci)

    if (!missing(obj)) b_sims <- b_sim(obj = obj, nsim = nsim)

    else if (missing(obj) & !missing(b_sims)) b_sims <- b_sims

    else if (missing(obj) & missing(b_sims) & !missing(mu) & !missing(Sigma))
        b_sims <- b_sim(mu = mu, Sigma = Sigma, nsim = nsim)

    if (missing(newdata) & !missing(obj))
        newdata <- find_scenarios(obj = obj, nsim = nsim,
                                 large_computation = large_computation)
    if (missing(newdata) & missing(obj))
        stop('At least one of obj and newdata must be supplied to find simulation scenarios.')

    qi_df <- linear_systematic(b_sims = b_sims, newdata = newdata, ...)

    if (missing(FUN)) {
        if (verbose)
            message('Note: FUN argument missing -> assuming b_sims is from a normal linear model.\n')
        names(qi_df)[grep('ls_', names(qi_df))] <- 'qi_'
    }
    else {
        FUN_check(FUN)

        temp_qi <- FUN(qi_df[['ls_']])

        FUN_results_check(temp_qi)

        qi_df[, 'qi_'] <- temp_qi
        qi_df['ls_'] <- NULL
    }

    if (ci < 1) {
        if (original_order) {
            orig_order <- unique(qi_df[, 1:(ncol(qi_df)-1)])
            orig_order <- apply(orig_order, 1, paste, collapse = '.')
        }

        qi_df$scenario_ <- interaction(qi_df[, 1:(ncol(qi_df)-1)])
        if (original_order)
            qi_df$scenario_ <- factor(qi_df$scenario_, levels = orig_order)

        qi_df <- qi_central_interval(qi_df, scenario_var = 'scenario_',
                                     qi_var= 'qi_', ci = ci)

        if (!isTRUE(slim)) qi_df$scenario_ <- NULL
    }

    if (slim) qi_df <- qi_slimmer(qi_df)

    return(qi_df)
}


R/linear_systematic.R
#' Find the systematic component in the linear form for fitted values in across
#' each simulation (note: largely for internal use by \code{\link{qi_builder}})
#'
#' @param b_sims a data frame created by \code{\link{b_sim}} of simulated
#' coefficients.
#' @param newdata a data frame of fitted values with column names corresponding
#' to variable names in \code{b_sims}. Variables in \code{b_sim} not present
#' in \code{newdata} will be treated as fitted at 0. Interactions will
#' automatically be found if they were entered into to the model using the
#' \code{*} operator.
#' @param inc_intercept logical whether to include the intercept in the
#' lineary systematic component.
#'
#' @return A data frame fitted values supplied in \code{newdata} and associated
#' linear systematic component estimates for all simulationed coefficient
#' estimates. The linear systematic components are included in a column
#' named \code{ls_}.
#'
#' @examples
#' library(car)
#'
#' # Estimate model
#' m1 <- lm(prestige ~ education + type, data = Prestige)
#'
#' # Create fitted values
#' fitted_df <- expand.grid(education = 6:16, typewc = 1)
#'
#' # Simulate coefficients
#' m1_sims <- b_sim(m1, nsim = 1000)
#'
#' # Find linear systematic component for fitted values
#' ls <- linear_systematic(b_sims = m1_sims, newdata = fitted_df)
#'
#' @source King, Gary, Michael Tomz, and Jason Wittenberg. 2000. "Making the
#' Most of Statistical Analyses: Improving Interpretation and Presentation."
#' American Journal of Political Science 44(2): 341-55.
#'
#' @importFrom dplyr bind_rows
#'
#' @export

linear_systematic <- function(b_sims, newdata, inc_intercept = TRUE)
{
    newdata <- non_linear_transformer(x = newdata, b_sims = b_sims)
    newdata <- factorise(x = newdata, b_sims = b_sims)

    fitted_names <- names(newdata)

    if (!('intercept_' %in% names(b_sims))) inc_intercept <- FALSE

    if (!all(fitted_names %in% names(b_sims)))
        stop('Unable to find all of the variables from newdata in b_sims.',
            call. = FALSE)

    if (!all(sapply(newdata, class) %in% c('numeric', 'integer')))
        stop('All fitted values must be either numeric or integer.',
            call. = FALSE)

    newdata <- interaction_builder(b_sims = b_sims, newdata = newdata)
    fitted_names <- names(newdata)

    intercept <- b_sims[['intercept_']]
    not_fitted_0 <- data.matrix(b_sims[, fitted_names])
    sims_fitted <- list(not_fitted_0, data.matrix(newdata))

    if (inc_intercept) {
        ls <- lapply(1:nrow(sims_fitted[[2]]), function(x) {
            fitted_1 <- sims_fitted[[2]][x, ]
            data.frame(
                data.frame(
                    t(fitted_1)),
                    ls_ = intercept + (sims_fitted[[1]] %*% fitted_1)
            )
        })
    }
    else {
        ls <- lapply(1:nrow(sims_fitted[[2]]), function(x) {
            fitted_1 <- sims_fitted[[2]][x, ]
            data.frame(
                data.frame(
                    t(fitted_1)),
                    ls_ = (sims_fitted[[1]] %*% fitted_1)
            )
        })
    }

    ls <- data.frame(bind_rows(ls))
    return(ls)
}
